### ISSUE TITLE
Bug in rebind report functionality

### ISSUE THREAD

**Issue by alxy** [codecutout/terraform-provider-powerbi#54]
The functionality that I introduced in #48 seems to have a bug. If we upload a report that is bound to a specific dataset in a different workspace already (which is the usual case, as reports always need a dataset set and for development activities the reports need to be connected to the development dataset), the rebinding to another dataset works fine when doing the initial upload. 

However, on further updates of the report, we see the following behavior:
![image](https://user-images.githubusercontent.com/2057062/141317826-3fbacad3-e1f7-4cc4-a673-e9fd38976352.png)

As you can see, the report is duplicated and bound to the original dataset in another workspace (that is the dataset the report is also linked to during development in PowerBI desktop). It looks like the rebinding doesn't happen again, and the report is duplicated (which seems to be PowerBIs default behavior if you upload a report with the same name to the workspace just bound to another dataset). The Terraform plan output however kind of looks correct (it shows a modification, not a re-upload):

```
An execution plan has been generated and is shown below.  
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # powerbi_pbix.example_report will be updated in-place
  ~ resource "powerbi_pbix" "example_report" {
        id                = "84042895-5c00-4de6-8ed6-3b2a5340103c"
        name              = "My report"
      ~ source_hash       = "27cf945c3439dfbfd8211c1e32d85614" -> "c5c73418721579f69a65b2cd2088104a"
        # (5 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

**Comment by alxy** [comment]
Found the main reason for this bug and it is as designed looking at the code: https://github.com/codecutout/terraform-provider-powerbi/blob/cda7bedc1fc0ff305cbcadb63fbea195d555d4e2/internal/powerbi/resource_pbix.go#L319-L332

However, as the updated report gets a new `report_id` (which is never reflected into the state) the rebind operation will still trigger a rebinding on the old report (first row in the screenshot above). I'm not sure what the original intention for this code was, can you shed some light onto this @alex-davies ?

**Comment by alxy** [comment]
I have implemented a first naive and incomplete fix in #55. The result is still something what I would consider broken/buggy, however, it would help us tremendously currently as having additional reports is not as bad and critical as having reports bound to the wrong dataset. (In our production environment currently all reports are bound to the dev datasets, with no easy way to rebind them. The situation is _really_ bad.)

The result with this fix applied is as follows:
![image](https://user-images.githubusercontent.com/2057062/141494336-09e0e4da-f8b2-4d89-8b4c-5ccc6c1fa70a.png)
For each additional update to the report we get a new report in the service portal.

**Comment by alex-davies** [comment]
The intention for the code was to keep `report_id` and `dataset_id` as the original IDs for those values when pbix was uploaded. 

Although I think it was assumed that rebinding only really affected the `dataset_id` and the id the report remained consistent. Can you indeed confirm that the behaviour of a rebind is to change the id of the report? Or is it another upload of the report that causes a duplicate with a different report_id?

PowerBI's behaviour with uplaoding pbix is not intuitive or clear enough to try and reason about theorteically. Can you get a failing integration test?

**Comment by alxy** [comment]
> Although I think it was assumed that rebinding only really affected the `dataset_id` and the id the report remained consistent. Can you indeed confirm that the behaviour of a rebind is to change the id of the report? Or is it another upload of the report that causes a duplicate with a different report_id?

That was also my thinking, however, if you upload a report that is originally bound to a different dataset (other than the one in the workspace you are currently deploying to, so in the example above `My dataset`) you get a completely new report and thus also a new `report_id`.

> PowerBI's behaviour with uplaoding pbix is not intuitive or clear enough to try and reason about theorteically. Can you get a failing integration test?

Providing such test is difficult, as you need at least two workspaces (one I'd call origin, where the original dataset is located, for example `PPC-KPI-Tool-Dataset` in the first example in a workspace called `dmp-dev` and the workspace you are deploying to, in my case typically also created by the provider). The report (PBIX) needs to be bound to that dataset in the origin workspace. Additionally, the service principal used to authenticate the provider needs access to both workspaces. I would assume that this is pretty hard if not impossible to arrange for an integration test.

However, I have now spent a couple of hours on this and do have a pretty good understanding what's going on. So if you like you can share some contact details with me and I can explain and/or demo in detail what is happening. We could then also discuss if it's possible to do that from an integration test :)

**Comment by alxy** [comment]
As an alternative solution to the problem at hand, I have just found the following. If the rebinding happens before the report is uploaded, a lot of the problems I have described earlier go away.

As this seems to be possible from looking at this [helper](https://github.com/codecutout/terraform-provider-powerbi/blob/master/internal/pbixrewriter/pbixPipelineFuncs.go), I'd propose that as an alternative fix to the bug. Not sure if that would have any downsides though.

**Comment by alex-davies** [comment]
@alxy  finally had some time to look more into this

That helper was built prior to the api that allowed rebinding. It was mostly an experiment and currently only used in tests to rebind to push datasets. It works by unzipping and altering the actual PBIX file and as such is entirely unsupported and we cant give any guarantees that it will continue to work with new features or updates to the pbix format

As for your current issue, my understanding of the issues is
* Imports will typically update both dataset and report in place (i.e. it will not assign new IDs)
* This update behaviour does not work if the report has been rebinded to a different dataset. In these cases it creates a new report with a new id
* Your 'fix' has its side affects because it forces a new report to be created everytime
* Mainting the ID of the report is critical otherwise every deploy will break everyones link to the report
* There is no way to force an import to update other than its default (and not very intuitive) behaviour

I believe I have a fix. By unbinding the dataset before the import it allows the import to perform its update, then we can rebind it back again

Would you be able to look at the changes on https://github.com/codecutout/terraform-provider-powerbi/pull/57/files and confirm my understanding (and confirm the fix) for the issue is correct

**Comment by alxy** [comment]
Your understanding of the issue is correct and your list nicely summarized it. 

However, I have tested your fix now locally and it does not seem to work, as the behavior I get is the same as I presented in the very first message in this issue. I'm not exactly sure why that as, as I think the bugfix makes sense:

- If there is an update to the report (which is rebound to a different dataset), first rebind the report to its original dataset
- Then, do the import (thus we don't get a new report and report_id and the import should correctly override)
- Lastly, rebind the report again to the given dataset in `rebind_dataset_id`

I will try to find out now why the fix is not working as intended.

**Comment by alxy** [comment]
OK, I guess I have found the issue. If I upload a report and do not use the `rebind_dataset_id` yet, I get the following result:

Terraform:
```terraform
resource "powerbi_pbix" "example_report" {
  workspace_id = powerbi_workspace.example.id
  name         = "My report"
  source       = "data/Reports/PCT_Production_One_Pager.pbix"
  source_hash  = filemd5("data/Reports/PCT_Production_One_Pager.pbix")
}
```

State:
```json
{
      "mode": "managed",
      "type": "powerbi_pbix",
      "name": "example_report",
      "provider": "provider[\"registry.terraform.io/codecutout/powerbi\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "dataset_id": null,
            "datasource": [],
            "id": "559b9d16-222c-4b9f-9ff6-9c4bd312e4d6",
            "name": "My report",
            "parameter": [],
            "rebind_dataset_id": null,
            "report_id": "d0e243c3-078d-4001-89c2-d504022072e4",
            "skip_report": false,
            "source": "data/Reports/PCT_Production_One_Pager.pbix",
            "source_hash": "964c9490149b324039711367020d3b7d",
            "timeouts": null,
            "workspace_id": "ce2347d6-6233-4998-92d0-ac2ce80b0f2b"
          },
          "sensitive_attributes": [],
        }
      ]
    },
```

Service portal:
![image](https://user-images.githubusercontent.com/2057062/142874023-66d9a48b-04e2-4f8a-8d8b-3b015f227f8f.png)

As you can see, `dataset_id` is `null` even though the report is bound to a dataset (that has been set in PowerBI desktop).

This means, later on the report can't be rebound to the original dataset (this should happen in your `unbindPBIXDataset()` function), because there is no reference to this original dataset anywhere in the state.

**Comment by alex-davies** [comment]
Interesting!

My mistake was assuming that the dataset deployed with the pbix was always going to be the dataset the report is attached to. That is obviously not the case for pbix files that have reports linking to an external dataset

In order not to complicate the existing `dataset_id` (which is defined as the dataset deployed in the pbix) I've introduced a new variable `report_original_dataset_id` solely to track the dataset the report is originally pointing to, for the sole purpose of unbinding

Although I don't like introducing new variable, i think it will be far messier to try and use other variables for this purpose

@alxy Take a look and play with the new PR and see if you can find any other use case I have missed

**Comment by alxy** [comment]
I have tested the fix now and it seems to be working fine for the use case we have. In particular, I have tested the following:

- Report updated: Result is no new `report_id` and the report is correctly updated
- Dataset updated: Dataset is updated, and the report is still bound to the correct dataset
- Remove the `rebind_datatset_id`: Report is bound back to its original dataset
- `terraform destroy`: All resources are properly removed

So to me this fix looks pretty good I have to say! Looking forward to test it properly in our production setup once it is released :)

**Comment by alxy** [comment]
@alex-davies Is there a chance to get this merged? :)

**Comment by cbogdanmihai** [comment]
@alxy @alex-davies Any news on this fix? It will be great to have it merged and in production, I'm building some huge pipelines for report deployments and this is a blocker for me, because the original report-id is lost when the report is duplicated after RebindReport action.

**Comment by alxy** [comment]
@cbogdanmihai The fix has been released already in #57 and is available with the latest version (v1.4.1) of this provider.
